rpc/src/services/stream_services.rs

```diff
@@ -70,163 +69,232 @@ pub struct MempoolOperationsQuery {
+                        reason: "Operation hash not found in operations data".into(),
+                    })
+                }
+            };
+            let monitored_op = MonitoredOperation {
+                branch: operation.branch().to_base58_check(),
+                protocol: Some(protocol_hash.to_string()),
+                hash: op_hash,
+                protocol_data: serde_json::from_str(&applied_op.protocol_data_json)?,
+                error: None,
+            };
+            result.push(monitored_op)
+        }
+        Ok(result)
+    }
+
+    pub fn collect_errored(
+        errored: &[Errored],
+        operations: &HashMap<OperationHash, Operation>,
+        protocol_hash: &str,
+        streamed_operations: &mut HashSet<String>,
+    ) -> Result<Vec<Self>, RpcServiceError> {
+        let mut result = Vec::with_capacity(errored.len());
+        for errored_op in errored {
+            let op_hash = errored_op.hash.to_base58_check();
+
+            // if the operation is allready included in the stream, ignore it
+            if streamed_operations.contains(&op_hash) {
+                continue;
+            }
+
+            streamed_operations.insert(op_hash.clone());
+
+            let operation = match operations.get(&errored_op.hash) {
+                Some(op) => op,
+                None => {
+                    return Err(RpcServiceError::NoDataFoundError {
+                        reason: "Operation hash not found in operations data".into(),
+                    })
+                }
+            };
+            let monitored_op = MonitoredOperation {
+                branch: operation.branch().to_base58_check(),
+                protocol: Some(protocol_hash.to_string()),
+                hash: op_hash,
+                protocol_data: serde_json::from_str(
+                    &errored_op
+                        .protocol_data_json_with_error_json
+                        .protocol_data_json,
+                )?,
+                error: Some(
+                    errored_op
+                        .protocol_data_json_with_error_json
+                        .error_json
+                        .clone(),
+                ),
+            };
+            result.push(monitored_op)
+        }
+        Ok(result)
+    }
 }
 
 pub struct HeadMonitorStream {
     block_meta_storage: BlockMetaStorage,
 
     state: RpcCollectedStateRef,
     last_checked_head: Option<BlockHash>,
-    delay: Option<Interval>,
     protocol: Option<ProtocolHash>,
+    contains_waker: bool,
+    stream_id: Uuid,
+    log: Logger,
 }
 
 pub struct OperationMonitorStream {
-    chain_id: ChainId,
+    _chain_id: ChainId,
     current_mempool_state_storage: CurrentMempoolStateStorageRef,
     state: RpcCollectedStateRef,
     last_checked_head: BlockHash,
     log: Logger,
-    delay: Option<Interval>,
-    streamed_operations: Option<HashSet<String>>,
+    contains_waker: bool,
+    stream_id: Uuid,
```

Comment:
@adonagy 
well, I would maybe do the wrapper here, so you dont need to spread `uuid` dependecy to rpc module :)
```
pub type StreamId = Uuid;

pub fn generate_stream_id() -> StreamId {
  Uuid::new_v4()
}

pub type StreamWakers = HashMap<StreamId, Waker>;
```

or maybe some wrapper, but then you need to reimpolement Eq,Hash,PartialEq...
```
pub struct StreamId(Uuid) {
   
}

impl StreamId {
   
pub fn generate() -> Self {..}
}

```

I guess the first version is enought

Reply:
I've gone with the your first suggestion. The trait would have been nice, but we would have to either:

1. lock the rwlock on the state that's implementing the trait
2. use a trait object to call `generate_stream_id` directly, this would add a dynamic dispatch the runtime (and could cause a small performance hit maybe)

Both cases add unnecessary complexity, so I went with the most simple wrapper approach :)